crates/chain-state/src/in_memory.rs

```diff
@@ -354,6 +356,27 @@ impl CanonicalInMemoryState {
     pub fn notify_canon_state(&self, event: CanonStateNotification) {
         self.inner.canon_state_notification_sender.send(event).ok();
     }
+
+    /// Return state provider with reference to in-memory blocks that overlay database state.
+    ///
+    /// This merges the state of all blocks that are part of the chain that the requested block is
+    /// the head of. This includes all blocks that connect back to the canonical block on disk.
+    pub fn state_provider(
+        &self,
+        hash: B256,
+        historical: StateProviderBox,
+    ) -> MemoryOverlayStateProvider {
+        let mut in_memory = Vec::new();
+        let mut current_state = self.state_by_hash(hash);
+
+        while let Some(state) = current_state {
+            in_memory.insert(0, state.block());
+
+            current_state = state.parent.as_ref().map(|parent| Arc::new(*parent.clone()));
+        }
```

Comment:
I'd like to convert this to two functions for convenience of BlockState that 
a) returns the parent blocks `Vec<BlockState>`
b) returns the entire chain: a) + push self